backup-stores/azure/src/main/java/io/camunda/zeebe/backup/azure/FileSetManager.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.1. You may not use this file
+ * except in compliance with the Zeebe Community License 1.1.
+ */
+package io.camunda.zeebe.backup.azure;
+
+import com.azure.core.util.BinaryData;
+import com.azure.core.util.Context;
+import com.azure.storage.blob.BlobClient;
+import com.azure.storage.blob.BlobContainerClient;
+import com.azure.storage.blob.models.BlobRequestConditions;
+import com.azure.storage.blob.models.BlobStorageException;
+import com.azure.storage.blob.options.BlobParallelUploadOptions;
+import io.camunda.zeebe.backup.api.BackupIdentifier;
+import io.camunda.zeebe.backup.api.NamedFileSet;
+import io.camunda.zeebe.backup.azure.AzureBackupStoreException.UnexpectedManifestState;
+import io.camunda.zeebe.backup.azure.shared.BlobRequestsOptionsManagement;
+import java.io.FileNotFoundException;
+import java.nio.file.NoSuchFileException;
+
+final class FileSetManager {
+
+  public static final int PRECONDITION_FAILED = 412;
+  // The path format is constructed by partitionId/checkpointId/nodeId/nameOfFile
+  private static final String PATH_FORMAT = "%s/%s/%s/%s/";
+  private final BlobContainerClient containerClient;
+  private boolean containerCreated = false;
+  private String fileSetPath;
+
+  FileSetManager(final BlobContainerClient containerClient) {
+    this.containerClient = containerClient;
+  }
+
+  void save(final BackupIdentifier id, final String fileSetName, final NamedFileSet fileSet)
+      throws NoSuchFileException {
+    for (final var namedFile : fileSet.namedFiles().entrySet()) {
+      final var fileName = namedFile.getKey();
+      final var filePath = namedFile.getValue();
+
+      if (!containerCreated) {
+        containerClient.createIfNotExists();
+        fileSetPath = fileSetPath(id, fileSetName);
+        containerCreated = true;
+      }
```

Comment:
:x: This is still not correct. I see that you added a test for this in `RestoringBackup`. :+1: For now, this test cannot run because restore is not implemented. We could add a weaker test that works with just `save`  and add it to `SavingBackup`. This would have caught the bug.

```java
  @ParameterizedTest
  @ArgumentsSource(TestBackupProvider.class)
  default void canSaveSecondBackup(final Backup firstBackup) {
    // given
    final Backup secondBackup =
        new BackupImpl(
            new BackupIdentifierImpl(
                firstBackup.id().nodeId() + 1,
                firstBackup.id().partitionId(),
                firstBackup.id().checkpointId()),
            firstBackup.descriptor(),
            firstBackup.snapshot(),
            firstBackup.segments());

    getStore().save(firstBackup).join();

    // when
    final var secondSave = getStore().save(secondBackup);

    // then
    assertThat(secondSave).succeedsWithin(Duration.ofMinutes(1));
  }
```

In this version, the test fails because it tries to overwrite a blob.

To fix this, let's think about what the FileSetManager has to do:
#### Create the container _once_ if it does not already exist
This can happen anywhere and the only thing we have to keep in mind is to not do it too often.
In the current version, you set a boolean flag to remember if you created the container yet. Then you check it for every file in a fileset which doesn't hurt but is unnecessary. I'd suggest to either do the `createIfNotExists` call once in the constructor of `FileSetManager` or `AzureBackupStore` or once at the beginning of the `FileSetManager#save` method. Doing it for every single file in  a fileset is not necessary.
#### Create a blob for every file in a fileset
Every blob has a name consisting of a fileset path which remains the same for every file in the fileset and the actual filename which is different for every file in the fileset.
Here you conflate this building of the constant fileset path with creating the container.
What we really want is to compute the fileset path once at the beginning of `FileSetManager#save` and not save it in an instance field because it is unique for every call to `FileSetManager#save`.